{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"view-in-github"},"source":["<a href=\"https://colab.research.google.com/github/boydvcu/VCU_BNFO301_Spring2022/blob/main/Lab5_Matrix_KEY_Student.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"]},{"cell_type":"markdown","metadata":{"id":"y_V08ygIMrJd"},"source":["<h1> Creating Consensus Sequences</h1>\n","\n","This program will create a consensus sequence, which represents the most frequent nucleotide at each position in an alignment of two or more sequences.  Consensus sequences can be useful for identifying and visualizing motifs in a DNA or amino acid sequence. The data is a FASTA file that contains sequences from Human, Gorilla, Dolphin, Rat and Mouse.\n","\n","> Indented block\n","\n","\n"]},{"cell_type":"markdown","metadata":{"id":"mMbh0oaoNIEm"},"source":["### Setup\n","Load the Sequence Data File.  Please run this block without changing the code.\n","\n","File : https://raw.githubusercontent.com/MusBansal/BNFO301Data/main/Assignment3_data.fasta"]},{"cell_type":"code","metadata":{"id":"TMJrCsgQMp2T","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1702751863164,"user_tz":300,"elapsed":260,"user":{"displayName":"Ashley Victor","userId":"03202414394000314274"}},"outputId":"f588a9c2-c187-49d2-8852-059e45f2f26a"},"source":["import os.path\n","# Load the genbank file\n","DATA_FILE_GITHUB = \"https://raw.githubusercontent.com/MusBansal/BNFO301Data/main/Assignment3_data.fasta\"\n","DEFAULT_FILE_NAME = 'Assignment3_data.fasta'\n","\n","fileName = DEFAULT_FILE_NAME\n","#Does the file exists locally, if not get it from the github\n","if not os.path.exists(fileName):\n","  #Load the file from Github to the local folder\n","  !wget --no-check-certificate --content-disposition $DATA_FILE_GITHUB\n","\n"],"execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["--2023-12-16 18:37:42--  https://raw.githubusercontent.com/MusBansal/BNFO301Data/main/Assignment3_data.fasta\n","Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.111.133, ...\n","Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 312 [text/plain]\n","Saving to: ‘Assignment3_data.fasta’\n","\n","\rAssignment3_data.fa   0%[                    ]       0  --.-KB/s               \rAssignment3_data.fa 100%[===================>]     312  --.-KB/s    in 0s      \n","\n","2023-12-16 18:37:42 (22.9 MB/s) - ‘Assignment3_data.fasta’ saved [312/312]\n","\n"]}]},{"cell_type":"markdown","metadata":{"id":"NgIx3OQrOCXT"},"source":["###Step 1.  Create a definition to clean the FASTA formated file and add sequences into a list.  \n","\n","We created a second list that contains the headers (with the \">\" removed).  \n","\n","We created function that can print the each list.  This later function will allow you to evaluate if your dictionary contains the expected information."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"2Vl6BmmKOWkV","outputId":"e8cf61fe-8eab-41d9-8b04-a4c14ddb26d9","executionInfo":{"status":"ok","timestamp":1702751867811,"user_tz":300,"elapsed":234,"user":{"displayName":"Ashley Victor","userId":"03202414394000314274"}}},"source":["#Read in fasta and create list of sequences (sequences) and a list of sequence identifiers (seqIDs)\n","\n","def ReadFile(fh):\n","    # Open file\n","    file = open(fh)\n","\n","    # Initialize variables\n","    sequences = []  # Store sequences\n","    seqIds = []  # Store sequence headers\n","\n","    # Loop through file\n","    for line in open(fh):\n","\n","        # Remove all unwanted white spaces\n","        line = line.strip()\n","\n","        # Find fasta header for first sequence\n","        if line.startswith(\">\"):\n","            seqIds.append(line.replace(\">\", \"\"))\n","        else:\n","            sequences.append(line)\n","\n","    # Return lists\n","    return sequences, seqIds\n","\n","\n","# -----------------------------\n","# Utility function to pretty print\n","# ------------------------------\n","def printInfo(seqIds, seqs):\n","    for seqId, seq in zip(seqIds, seqs):\n","        print(\"{: <12} {: <20}\".format(seqId, seq))\n","\n","\n","sequences, seqIds = ReadFile(fileName)\n","# print(sequences, seqIds)\n","printInfo(seqIds, sequences)\n","\n","\n","\n"],"execution_count":2,"outputs":[{"output_type":"stream","name":"stdout","text":[" Human       GGAGAGGCTCGGAGCCGGGCCCGGACCCCGGCGATTGCCGCCCGCTTCTCT\n"," Gorilla     GGAGAGGCTCGGAGCCGGGCCCGGACCCCGGCGATTGCCGCCCGCTTCTCT\n"," Dolphin     GAGGCTC---GGAGCCGGACCTGGACCCCTGCGATAGCCGTCTG-CTCCCG\n"," Rat         GGAGCAACTAGGAACCCGAACCAGAGCCCGGCGAGCGCAGCCTGCAGCTCC\n"," Mouse       GAGGCGCCTAGGAACCCGAGCCGGAGCTCAGCGAGCGCAGCCTGCAGCTCC\n"]}]},{"cell_type":"code","metadata":{"id":"wrrprj_p_sYT","outputId":"6329cf79-64c3-433a-a4c9-6360b8e132c8","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1702752052855,"user_tz":300,"elapsed":8,"user":{"displayName":"Ashley Victor","userId":"03202414394000314274"}}},"source":["\n","# -----------------------------\n","# Convert sequence to matrix for easy computation\n","# ------------------------------\n","def generateMatrix(seqs):\n","    # Lengths of the aligned sequences are equal\n","    # Get length of first sequence\n","    length = len(seqs[0])\n","\n","    # Initialize variables\n","    matrix = []  # Will store all values at each position\n","\n","    # Get a list of list\n","    # Reads fasta file as matrix\n","    # Example:\n","    #        ATGCA\n","    #        ATGAA\n","    #        TCGAT\n","    #             Bases at index 0   Bases at index 1 ...\n","    # positions = [[\"A\", \"A\", \"T\"],  [\"T\", \"T\", \"C\"], ...]]\n","\n","    for base in range(length):\n","        column = []\n","        for sequence in range(len(seqs)):\n","            column.append(seqs[sequence][base])\n","        matrix.append(column)\n","\n","    return matrix\n","\n","\n","matrix = generateMatrix(sequences)\n","\n","#printing first 3 rows to show the matrix\n","print(matrix[0:3])\n","\n"],"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["[['G', 'G', 'G', 'G', 'G'], ['G', 'G', 'A', 'G', 'A'], ['A', 'A', 'G', 'A', 'G']]\n"]}]},{"cell_type":"markdown","metadata":{"id":"4E2GsDEt_8A7"},"source":["###Step-2.  Creating a function that returns a consensus sequence using the most fequent base.\n","\n","1. If all bases are in agreement, return base as upper.case.  e.g. ['G', 'G', 'G', 'G', 'G'] returns G\n","2. If there is an equal split between bases in a column, return \"n\".  e.g. ['G', 'G', 'A', 'A', 'A', 'G'] returns n\n","3. If the the bases are not in agreement, but one base is more frequent than other, return the most frequent base in lower.case.  eg. ['G', 'G', 'G', 'G', 'A'] returns g\n","\n","The output should look like this: GgaGcg?ct?GGAgCCgGacCcgGAcCcCgGCGAt?GCcGcCtGc?tCtC?\n"]},{"cell_type":"code","metadata":{"id":"lx3sHnCK_-LO","outputId":"0833f93f-c66e-45d4-dff9-903f7d673675","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1702752057085,"user_tz":300,"elapsed":744,"user":{"displayName":"Ashley Victor","userId":"03202414394000314274"}}},"source":["def getConsensus(matrix):\n","    # Initialize variables\n","    consensus = \"\"\n","\n","    # Loop through each inner list\n","    for row in matrix:\n","        freq = {}\n","        for char in row:\n","            # Get counts for each base\n","            freq[char] = freq.get(char, 0) + 1\n","\n","        # Get the max value\n","        maxFreq = max(freq, key=freq.get)\n","\n","        # See if there are multiple keys with the max value\n","        maxKeys = [k for k, v in freq.items() if v == freq.get(maxFreq)]\n","\n","        # Calculate abundance of base at position\n","        abundance = float(freq.get(maxFreq)) / len(row)\n","        if abundance == 1:\n","            consensus += maxKeys[0].upper()\n","        elif len(maxKeys) == 1 and abundance > 0.5:  # Base is common at position\n","            consensus += maxKeys[0].lower()\n","        else:\n","            consensus += \"?\"\n","\n","    return consensus\n","\n","\n","consensus = getConsensus(matrix)\n","print(\"{: <25} {: <20}\\n\".format(\"CONSENSUS\", consensus))\n"],"execution_count":4,"outputs":[{"output_type":"stream","name":"stdout","text":["CONSENSUS                 GgaGcg?ct?GGAgCCgGacCcgGAcCcCgGCGAt?GCcGcCtGc?tCtC?\n","\n"]}]},{"cell_type":"markdown","metadata":{"id":"A1z6jSbwCxTs"},"source":["###Bonus  \n","One way aligned sequences are often visualized is as a sequence logo. For an additional point, go to https://weblogo.berkeley.edu/logo.cgi and upload the provided FASTA file. Save the image and upload to canvas.  You may do the bonus between class meetings on your own time."]}]}