{"nbformat":4,"nbformat_minor":0,"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.7.4"},"colab":{"provenance":[]}},"cells":[{"cell_type":"markdown","metadata":{"id":"view-in-github"},"source":["<a href=\"https://colab.research.google.com/github/boydvcu/VCU_BNFO301_Spring2022/blob/main/KEY_BNFO_301_regex_part2.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"]},{"cell_type":"markdown","metadata":{"id":"CV8ltn7IuWLR"},"source":["<h1>Finding Degenerate Primers using Regex</h1>\n","\n","This python program finds degenerate primers using regex."]},{"cell_type":"markdown","metadata":{"id":"CM7DEBptKwuv"},"source":["### Part 0- Setup (Provided code)\n","Read the Genbank Sample Data File"]},{"cell_type":"code","metadata":{"id":"gjQcjXJLuWLV","outputId":"2c4f2ff4-b876-4e7a-9d2f-ad162d6291b1","colab":{"base_uri":"https://localhost:8080/"}},"source":["import re # python regex module\n","import os.path\n","\n","#Read the sequence file\n","DATA_FILE_GITHUB = \"https://raw.githubusercontent.com/MusBansal/BNFO301Data/main/primersequencewhole.txt\"\n","DEFAULT_FILE_NAME = 'primersequencewhole.txt'\n","\n","fileName = DEFAULT_FILE_NAME\n","#Does the file exists locally, if not get it from the github\n","if not os.path.exists(fileName):\n","  #Load the file from Github to the local folder\n","  !wget --no-check-certificate --content-disposition $DATA_FILE_GITHUB\n","\n","print(\"Reading file:\", fileName)"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["--2022-02-17 11:13:34--  https://raw.githubusercontent.com/MusBansal/BNFO301Data/main/primersequencewhole.txt\n","Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n","Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 582665 (569K) [text/plain]\n","Saving to: ‘primersequencewhole.txt’\n","\n","primersequencewhole 100%[===================>] 569.01K  --.-KB/s    in 0.03s   \n","\n","2022-02-17 11:13:34 (17.9 MB/s) - ‘primersequencewhole.txt’ saved [582665/582665]\n","\n","Reading file: primersequencewhole.txt\n"]}]},{"cell_type":"markdown","metadata":{"id":"wh0WbFnMqadD"},"source":["### Testing degenerative primers: <br>\n","<br>\n","PCR is a commonly used tool to isolate short regions of a genome.  PCR requires the use of primers, short sequences of DNA that bind to specific sites in a genome.  It is important to test primers, to determine if they will bind to more than one site in a genome.  Here you will use regex to identify matches to each of the four primers.  Please note that the primers are degenerate, which means that the primer will be mixed (e.g. some primers may have an A at a given site, while the others of a C)  You will be provided with a fasta file.  Please examine this file carefully to determine if you need to remove new line characters before identifying primer matches.\n","\n","* M-1B, 5′-agagtttgatmcacc-3′ (M -> A or C)\n","* M-2B, 5′-ctgctgcsycc-3′ (Y -> C or T), (S -> G or C)\n","* M-3B, 5′-gcaacgaa3′\n","* M-4B, 5′-ggcggtgtgtrc-3′ (R->A or G)\n","\n","<br>"]},{"cell_type":"code","metadata":{"id":"WZspvd_KypM9","outputId":"5f371d4d-9af8-487c-dd99-c376cca9f01a","colab":{"base_uri":"https://localhost:8080/"}},"source":["#Read the sequence file\n","#DATA_FILE_GITHUB = \"https://raw.githubusercontent.com/MusBansal/BNFO301Data/main/primersequencewhole.txt\"\n","#DEFAULT_FILE_NAME = 'primersequencewhole.txt'\n","\n","#fileName = DEFAULT_FILE_NAME\n","#Does the file exists locally, if not get it from the github\n","#if not os.path.exists(fileName):\n","  #Load the file from Github to the local folder\n","#  !wget --no-check-certificate --content-disposition $DATA_FILE_GITHUB\n","\n","#print(\"Reading file:\", fileName)\n","\n","#function to remove header lines using regex (\"that starts with >\")\n","######################\n","def removeHeaders(data):\n","  pattern = \"^>(.)*(\\n)*\"\n","  # Compile our search string\n","  positive_residues = re.compile(pattern)\n","\n","  # Get a list of all the segments\n","  list_of_seq_segments = positive_residues.findall(gb_input)\n","  print(\"Total Sequences Found for Replacement:\",len(list_of_seq_segments))\n","\n","  # Find and replace\n","  output = positive_residues.sub('', gb_input)\n","  return output\n","######################\n","\n","\n","# Read in the string from the file\n","with open(fileName, \"r\") as myfile:\n","    gb_input = myfile.read()\n","\n","gb_input = removeHeaders(gb_input)\n","\n","\n","# #Saving it in output file\n","# OUTPUT_FILE_NAME = 'HW2_Results1.txt'\n","# outfile = open(OUTPUT_FILE_NAME, \"w\")\n","# print(\"Writing to the file:\", OUTPUT_FILE_NAME)\n","# print(output, file=outfile)\n","# # close the output file\n","# outfile.close()\n","\n","\n","\n","#Since the file contains newlines, remove all the newlines\n","#this pattern will look for all the newline/tabs etc.\n","pattern = \"(\\r\\n)+|\\r+|\\n+|\\t+\"\n","\n","# Compile our search string\n","positive_residues = re.compile(pattern)\n","\n","\n","# Get a list of all the segments\n","list_of_seq_segments = positive_residues.findall(gb_input)\n","print(\"Total Sequences Found for Replacement:\",len(list_of_seq_segments))\n","\n","# Find and replace\n","output = positive_residues.sub('', gb_input)\n","#print(output)\n","\n","\n","\n","#function to find sequences\n","def findSequences(pattern,data):\n","  count = 0\n","  for match in re.finditer(pattern, data):\n","    count = count + 1\n","    # Start index of match (integer)\n","    sStart = match.start()\n","\n","    # Final index of match (integer)\n","    sEnd = match.end()\n","\n","    # Complete match (string)\n","    sGroup = match.group()\n","\n","    # Print match\n","    print('Match: {} \"{}\" found at: [{},{}]'.format(count, sGroup, sStart,sEnd))\n","  print(\"Total Sequences Found:\",count, \"\\n\")\n","  return count\n","\n","# Finding the patterns\n","\n","#1. Finding the pattern \"agagtttgatmcacc\"\n","#(?i) - ignores the case, m can be a or c -> [ac]\n","pattern = \"(?i)agagtttgat[ac]cacc\"\n","findSequences(pattern,output)\n","\n","#2. Finding the pattern \"ctgctgcsycc\"\n","#y can be c or t -> [ct]\n","pattern = \"(?i)ctgctg[gc][ct]cc\"\n","findSequences(pattern,output)\n","\n","#3. Finding the pattern \"gcaacgaa\"\n","pattern = \"(?i)gcaacgaa\"\n","findSequences(pattern,output)\n","\n","#4. Finding the pattern \"(?i)ggcggtgtgt[ag]c\"\n","#r can be a or g -> [ag]\n","pattern = \"(?i)ggcggtgtgt[ag]c\"\n","findSequences(pattern,output)\n","\n"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Total Sequences Found for Replacement: 1\n","Total Sequences Found for Replacement: 8206\n","Match: 1 \"AGAGTTTGATACACC\" found at: [5418,5433]\n","Total Sequences Found: 1 \n","\n","Match: 1 \"CTGCTGGTCC\" found at: [26006,26016]\n","Match: 2 \"CTGCTGGTCC\" found at: [559709,559719]\n","Total Sequences Found: 2 \n","\n","Match: 1 \"GCAACGAA\" found at: [73955,73963]\n","Match: 2 \"GCAACGAA\" found at: [92807,92815]\n","Match: 3 \"GCAACGAA\" found at: [162554,162562]\n","Match: 4 \"GCAACGAA\" found at: [192590,192598]\n","Match: 5 \"GCAACGAA\" found at: [482871,482879]\n","Total Sequences Found: 5 \n","\n","Match: 1 \"GGCGGTGTGTAC\" found at: [38814,38826]\n","Match: 2 \"GGCGGTGTGTAC\" found at: [337431,337443]\n","Total Sequences Found: 2 \n","\n"]},{"output_type":"execute_result","data":{"text/plain":["2"]},"metadata":{},"execution_count":2}]}]}